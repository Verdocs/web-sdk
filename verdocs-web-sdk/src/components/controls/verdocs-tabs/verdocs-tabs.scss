@import '../../../theme.scss';

verdocs-tabs {
  gap: 10px;
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  font-family: $primary-font;

  .tab {
    display: flex;
    color: #111111;
    cursor: pointer;
    font-size: 14px;
    font-weight: 400;
    padding: 5px 10px;
    align-items: center;
    text-transform: none;
    justify-content: center;
    border-bottom: 4px solid transparent;

    &:focus {
      border: none;
      outline: none;
    }

    &.active {
      color: #000000;
      font-weight: 500;
      border-bottom: 4px solid #654dcb;
    }

    &.disabled {
      color: #666666;
      cursor: default;
    }
  }
}
