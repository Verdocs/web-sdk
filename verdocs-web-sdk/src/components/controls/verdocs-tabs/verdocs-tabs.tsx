import {Component, Prop, Host, h, Event, EventEmitter} from '@stencil/core';

export interface ITab {
  id?: string;
  label: string;
  disabled?: boolean;
  // TODO: Tooltips?
}

/**
 * Display a simple row of selectable tabs. This is a controlled element.
 * The parent must adjust selectedTab as selection events are fired.
 */
@Component({
  tag: 'verdocs-tabs',
  styleUrl: 'verdocs-tabs.scss',
})
export class VerdocsTabs {
  /**
   * The tabs to display
   */
  @Prop() tabs: ITab[] = [];

  /**
   * The index of the tab to show selected.
   */
  @Prop() selectedTab: number = 0;

  /**
   * Event fired when the user clicks a template to view it. Typically the host application will use this to navigate
   * to the template preview. This is also fired when the user selects "Preview/Send" fropm the dropdown menu.
   */
  @Event({composed: true}) selectTab: EventEmitter<{tab: ITab; index: number}>;

  handleSelectTab(index: number) {
    this.selectTab?.emit({tab: this.tabs[index], index});
  }

  render() {
    return (
      <Host>
        {this.tabs.map((tab, index) => (
          <div class={`tab ${index === this.selectedTab ? 'active' : ''} ${tab.disabled ? 'disabled' : ''}`} onClick={tab.disabled ? () => {} : () => this.handleSelectTab(index)}>
            {tab.label}
          </div>
        ))}

        <div class="flex flex-1" />

        <slot>
          <div />
        </slot>
      </Host>
    );
  }
}
