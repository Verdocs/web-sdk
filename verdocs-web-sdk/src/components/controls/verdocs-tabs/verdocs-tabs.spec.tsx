import { newSpecPage } from '@stencil/core/testing';
import { VerdocsTabs } from './verdocs-tabs';
import { h } from '@stencil/core';

describe('verdocs-tabs', () => {
  beforeAll(() => {
    jest.spyOn(Math, 'random').mockReturnValue(0.123456789);
  });

  afterAll(() => {
    // restore the original behavior after the tests
    (Math.random as jest.Mock).mockRestore();
  });

  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsTabs],
      template: () => <verdocs-tabs tabs={[]} selectedTab={0}></verdocs-tabs>,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with tabs and selectedTab', async () => {
    const tabs = [
      { label: 'Tab 1' },
      { label: 'Tab 2', disabled: true }
    ];
    const page = await newSpecPage({
      components: [VerdocsTabs],
      template: () =>  <verdocs-tabs selectedTab={1} tabs={tabs}></verdocs-tabs>,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('emits selectTab event on tab click', async () => {
    const tabs = [
      { label: 'Tab 1' },
      { label: 'Tab 2' }
    ];
    const page = await newSpecPage({
      components: [VerdocsTabs],
      template: () =>  <verdocs-tabs selectedTab={1} tabs={tabs}></verdocs-tabs>,
    });

    const spy = jest.fn();
    page.root.addEventListener('selectTab', spy);

    const tabEls = page.root.querySelectorAll('.tab');
    if (tabEls.length > 1) {
      tabEls[1].dispatchEvent(new Event('click', { bubbles: true }));
      expect(spy).toHaveBeenCalled();
    }
  });
});
