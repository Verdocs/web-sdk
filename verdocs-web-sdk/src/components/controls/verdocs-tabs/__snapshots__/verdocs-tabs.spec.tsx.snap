// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-tabs renders with default props 1`] = `
<verdocs-tabs>
  <!---->
  <div class="flex flex-1"></div>
  <slot-fb>
    <div></div>
  </slot-fb>
</verdocs-tabs>
`;

exports[`verdocs-tabs renders with tabs and selectedTab 1`] = `
<verdocs-tabs>
  <!---->
  <div class="tab">
    Tab 1
  </div>
  <div class="active disabled tab">
    Tab 2
  </div>
  <div class="flex flex-1"></div>
  <slot-fb>
    <div></div>
  </slot-fb>
</verdocs-tabs>
`;
